### PR TITLE
FTS: Adds Full Text Search and Indexing Policy

### PR DESCRIPTION
**Pull request by aayush3011**
# Description

This PR adds a new `FullTextPolicy ` in `ContainerProperties` and a new `FullTextIndexes` in  the `IndexingPolicy` to enable Full Text Search in Cosmos DB.

Below is a sample for the new `ContainerProperties` with Full Text Indexing Policy andFull Text Policy highlighted.

```
{
  "id": "e1711280-b632-4d14-934f-5476d11f2472",
  "indexingPolicy": {
    "indexingMode": "consistent",
    "automatic": true,
    "includedPaths": [
      {
        "path": "/name/?"
      },
      {
        "path": "/description/?"
      }
    ],
    "excludedPaths": [
      {
        "path": "/*"
      },
      {
        "path": "/\"_etag\"/?"
      }
    ],
    "fullTextIndexes": [
      {
        "path": "/fts1"
      }
    ]
  },
  "partitionKey": {
    "paths": [
      "/mypk"
    ],
    "kind": "Hash"
  },
  "uniqueKeyPolicy": {
    "uniqueKeys": []
  },
  "conflictResolutionPolicy": {
    "mode": "LastWriterWins",
    "conflictResolutionPath": "/_ts",
    "conflictResolutionProcedure": ""
  },
  "geospatialConfig": {
    "type": "Geography"
  },
  "fullTextPolicy": {
    "defaultLanguage": "en-US",
    "fullTextPaths": [
      {
        "path": "/fts1",
        "language": "en-US"
      }
    ]
  },
  "_rid": "7loJAMpeA+k=",
  "_ts": 1730776275,
  "_self": "dbs/7loJAA==/colls/7loJAMpeA+k=/",
  "_etag": "\"00000000-0000-0000-2f30-5fd2d43501db\"",
  "_docs": "docs/",
  "_sprocs": "sprocs/",
  "_triggers": "triggers/",
  "_udfs": "udfs/",
  "_conflicts": "conflicts/"
}

```

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Path settings within {@link CosmosFullTextPolicy}
+ */
+public class CosmosFullTextPath {
+    @JsonProperty(Constants.Properties.LANGUAGE)
+    private String path;
+    @JsonProperty(Constants.Properties.PATH)
+    private String language;
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java:18]: The annotations are reversed.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        database = createDatabase(client, databaseId);
+    }
+
+    @AfterClass(groups = {"emulator"}, timeOut = SHUTDOWN_TIMEOUT, alwaysRun = true)
+    public void afterClass() {
+        safeDeleteDatabase(database);
+        safeClose(client);
+    }
+
+    @Test(groups = {"emulator"}, timeOut = TIMEOUT*10000)
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java:71]: 30_000 * 10_000 - that seems like a very large timeout.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        CosmosContainerProperties collectionDefinition = new CosmosContainerProperties(UUID.randomUUID().toString(), partitionKeyDef);
+
+        IndexingPolicy indexingPolicy = new IndexingPolicy();
+        indexingPolicy.setIndexingMode(IndexingMode.CONSISTENT);
+        ExcludedPath excludedPath = new ExcludedPath("/*");
+        indexingPolicy.setExcludedPaths(Collections.singletonList(excludedPath));
+
+        IncludedPath includedPath1 = new IncludedPath("/name/?");
+        IncludedPath includedPath2 = new IncludedPath("/description/?");
+        indexingPolicy.setIncludedPaths(ImmutableList.of(includedPath1, includedPath2));
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java:87]: Create a helper method to eliminate some duplication of this boilerplate in the below test?

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        CosmosFullTextIndex cosmosFullTextIndex = new CosmosFullTextIndex("/fts1");
+        indexingPolicy.setCosmosFullTextIndexes(Arrays.asList(cosmosFullTextIndex));
+
+        CosmosFullTextPath cosmosFullTextPath = new CosmosFullTextPath();
+        cosmosFullTextPath.setPath("/fts1");
+        cosmosFullTextPath.setLanguage("en-US");
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java:94]: There is no point in specifying this path if it's the same language as the default, right?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Constants.java [hunk 5]
```diff
@@ -120,6 +120,18 @@ public static final class Properties {
         public static final String SPATIAL_INDEXES = "spatialIndexes";
         public static final String TYPES = "types";
 
+        // Full text search
+        public static final String FULL_TEXT_SEARCH_PATHS = "fullTextIndexes";
+        public static final String GLOBAL_STATISTICS_QUERY_INFO = "globalStatisticsQueryInfo";
+        public static final String COMPONENT_QUERY_INFOS = "componentQueryInfos";
+        public static final String PROJECTION_QUERY_INFO = "projectionQueryInfo";
+        public static final String SKIP = "skip";
+        public static final String TAKE = "take";
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Constants.java:129]: These aren't necessary for this PR, right?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param path the path for the cosmosFullText.
+     * @return CosmosFullTextPath
+     */
+    public CosmosFullTextPath setPath(String path) {
+        if (StringUtils.isEmpty(path)) {
+            throw new NullPointerException("full text search path is either null or empty");
+        }
+
+        if (path.charAt(0) != '/' || path.lastIndexOf('/') != 0) {
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java:46]: So we don't allow nested paths to have a language other than the default?  Is that something worth validating in the client?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return defaultLanguage;
+    }
+
+    /**
+     * Sets the default language for cosmosFullText.
+     * @param defaultLanguage the default language for cosmosFullText.
+     * @return CosmosFullTextPolicy
+     */
+    public CosmosFullTextPolicy setDefaultLanguage(String defaultLanguage) {
+        this.defaultLanguage = defaultLanguage;
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java:42]: Doesn't this need to set it in the `jsonSerializable`?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Path settings within {@link CosmosFullTextPolicy}
+ */
+public class CosmosFullTextPath {
+    @JsonProperty(Constants.Properties.PATH)
+    private String path;
+    @JsonProperty(Constants.Properties.LANGUAGE)
+    private String language;
+    private JsonSerializable jsonSerializable;
```
Comment (Pilchie) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java:19]: This seems to be unused (Note, I'm not well-versed in how serialization works in Java).

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java [hunk 9]
```diff
@@ -441,6 +443,33 @@ public void setVectorEmbeddingPolicy(CosmosVectorEmbeddingPolicy value) {
...
+        return this.cosmosFullTextPolicy;
+    }
+
+    /**
+     * Gets the Full Test Policy containing paths for full text search and the language specification for each path.
+     * It also contains the default language to be used.
+     *
+     * @param value the FullTextPolicy
+     */
+    public void setFullTextPolicy(CosmosFullTextPolicy value) {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:468]: Its an existing pattern it-seems. 

Set -> Get -> Set -> Get
Lat Get will not see the latest set right? Is that okey?

Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:468]: Its internal not public contract, based on usage may be okey

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java (hunk 6 above)
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java:46]: At-least in .NET nested paths are allowed. 
Again IMO we can leave client implementation some what flexible and delegate such to the service (might help on future proofing as well)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java (hunk 7 above)
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java:42]: It seems It's a pure POJO not wrapping jsonserializable like others but using JsonAnnotations.
Would be a good learning point on when to use what.

**Comment by kirankumarkolli** [comment]
Can you please list the JAVA usage examples also in the description.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/IndexingPolicy.java [hunk 10]
```diff
@@ -310,6 +311,35 @@ public IndexingPolicy setVectorIndexes(List<CosmosVectorIndexSpec> vectorIndexes
...
+    public List<CosmosFullTextIndex> getCosmosFullTextIndexes() {
+        if (this.cosmosFullTextIndexes == null) {
+            this.cosmosFullTextIndexes = this.jsonSerializable.getList(Constants.Properties.FULL_TEXT_SEARCH_PATHS, CosmosFullTextIndex.class);
+
+            if (this.cosmosFullTextIndexes == null) {
+                this.cosmosFullTextIndexes = new ArrayList<CosmosFullTextIndex>();
+            }
+        }
+
+        return this.cosmosFullTextIndexes;
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/IndexingPolicy.java:328]: We should return an `Unmodifiable` copy of this.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java (hunk 2 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java:71]: Removed the 10000

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java (hunk 7 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java:42]: In the Java SDK we are using jsonSerializable in the top POJO, and using JsonAnnotations for the inner POJO's. Just following what is been followed across the SDK.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java (hunk 6 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java]: Fixed this, nested paths are allowed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java (hunk 9 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:468]: This pattern was implemented when the SDK was ported from V2 to V4.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java [hunk 11]
```diff
@@ -369,6 +369,28 @@ public CosmosContainerProperties setVectorEmbeddingPolicy(CosmosVectorEmbeddingP
...
+     * It also contains the default language to be used.
+     *
+     * @return the FullTextPolicy
+     */
+    public CosmosFullTextPolicy getFullTextPolicy() {
+        return this.documentCollection.getFullTextPolicy();
+    }
+
+    /**
+     * Gets the Full Test Policy containing paths for full text search and the language specification for each path.
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java:383]:
```suggestion
     * Gets the Full Text Policy containing paths for full text search and the language specification for each path.
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextIndex.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+package com.azure.cosmos.models;
+
+import com.azure.cosmos.CosmosItemSerializer;
+import com.azure.cosmos.implementation.Constants;
+import com.azure.cosmos.implementation.JsonSerializable;
+
+/**
+ * Represents cosmos full text index of the IndexingPolicy in the Azure Cosmos DB database service.
+ */
+public class CosmosFullTextIndex {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextIndex.java:13]:
```suggestion
public final class CosmosFullTextIndex {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+import com.azure.cosmos.implementation.Constants;
+import com.azure.cosmos.implementation.JsonSerializable;
+import com.azure.cosmos.implementation.apachecommons.lang.StringUtils;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Path settings within {@link CosmosFullTextPolicy}
+ */
+public class CosmosFullTextPath {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java:14]:
```suggestion
public final class CosmosFullTextPath {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+/**
+ * Full Text Search Policy
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CosmosFullTextPolicy {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java:16]:
```suggestion
public final class CosmosFullTextPolicy {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    @JsonProperty(Constants.Properties.DEFAULT_LANGUAGE)
+    private String defaultLanguage;
+    @JsonProperty(Constants.Properties.FULL_TEXT_PATHS)
+    private List<CosmosFullTextPath> paths;
+//    private JsonSerializable jsonSerializable;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java:21]: We should remove it if its not required anymore.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+    protected static final int TIMEOUT = 30000;
+    protected static final int SETUP_TIMEOUT = 20000;
+    protected static final int SHUTDOWN_TIMEOUT = 20000;
+    private static final String PATH = "/fts1";
+    private static final String NESTED_PATH = "/fts1/fts2";
+    private static final String LANGUAGE = "en-US";
+    private static final String DEFAULT_LANGUAGE = "en-US";
+
+
+    protected static Logger logger = LoggerFactory.getLogger(VectorIndexTest.class.getSimpleName());
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java]: Updated

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Ignore;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.UUID;
+
+@Ignore("TODO: Ignore these test cases until the public emulator with vector indexes is released.")
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java (hunk 11 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java (hunk 13 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextIndex.java (hunk 12 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextIndex.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java (hunk 14 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java (hunk 15 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPolicy.java]: Updated

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+
+    protected static Logger logger = LoggerFactory.getLogger(VectorIndexTest.class.getSimpleName());
+    private final ObjectMapper simpleObjectMapper = Utils.getSimpleObjectMapper();
+    private final String databaseId = CosmosDatabaseForTest.generateId();
+    private CosmosAsyncClient client;
+    private CosmosAsyncDatabase database;
+
+    @BeforeClass(groups = {"emulator"}, timeOut = SETUP_TIMEOUT)
+    public void before_VectorIndexTest() {
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java]: Updated

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/FullTextIndexTest.java]: AFAIK it does not has those changes. I'll confirm from Aditya once.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java (hunk 8 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosFullTextPath.java]: Removed it, as it was not being used anywhere.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java [hunk 19]
```diff
@@ -441,6 +444,35 @@ public void setVectorEmbeddingPolicy(CosmosVectorEmbeddingPolicy value) {
         this.set(Constants.Properties.VECTOR_EMBEDDING_POLICY, value, CosmosItemSerializer.DEFAULT_SERIALIZER);
     }
 
+    /**
+     * Gets the Full Text Policy containing paths for full text search and the language specification for each path.
+     * It also contains the default language to be used.
+     *
+     * @return the FullTextPolicy
+     */
+    @Beta(value = Beta.SinceVersion.V4_65_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:453]: This is implementation class, doesn't need Beta tag :)
please remove it.

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @aayush3011

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java (hunk 19 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:453]: Removed it.

**Pull request merged**